### DISCUSSION TITLE
Proposal : service selector for blueprints

### DISCUSSION THREAD

**Discussion by EPMatt** [home-assistant/architecture#506] [Entity Models] [answered]
Hi there!

I'd like to propose a new selector for blueprints:
* `service` for picking a service from a list of available ones.
```yaml
my_input:
  selector:
    service:
      integration: deconz
```

In the UI the input could appear in the form of the system service picker found in *Developer Tools -> Services*.
The `integration` option allows to filter services based on the integration which they are exposed by.

One use case of this input could be the user's need to specify a service to run inside the automation, but not an action.

For example, having the service selector would allow the user to select the desired `notify.mobile_app_<your_device_id_here>` when building a blueprint for sending notifications to devices. Users should not compile service data by themselves, since they can be provided by the blueprint (eg. when asking to pick a service from a known "family" of services which use the same data format).
 
I was working on a blueprint which could benefit from this feature, but I hope this could be helpful in general.

This proposal was moved from: https://github.com/home-assistant/architecture/issues/486

**Comment by mdegat01** [comment]
This is a bit trickier then this isn't it? If you just want any service then I think you should just use the action selector. When I think this selector comes up is when you want any service built for a particular platform with known input signature.

The most common example I think is notification services. There are many many integrations which include the notification platform. All of these then create one or more `notify.<something>` services which at a minimum accept `title`, `message` and `target`. I want to make a blueprint that allows users to pick from any of the notification services their various integrations have created but let my blueprint fill in the values for `title` and/or `message` since I know that service will have those inputs. The user should also be able to include additional values for `target` and any integration specific options that particular notification service has like `data`. I actually did this [here](https://community.home-assistant.io/t/get-notified-any-time-there-is-a-new-home-assistant-alert/256924) but its really ugly without a proper selector.

Although I have to admit, its hard to think of examples where you need this besides notifications. Normally you would just ask users for devices of a particular type that have known action options, like lights or a vacuum cleaner. There's just this gap when there's no physical thing to ask about. I want to prompt users for a thing which has a `notify` service, regardless of what that thing is (or if it even is a physical thing).

**Comment by frenck** [comment]
The example given for notification services, is wrong IMHO. It would make more sense to use a selector for notifications in that case.

For mobile_app specifically, this can already be done by using device automation (as is used by the default blueprints shipped with HA itself as well).

Services are always part of an action or sequence, for which we already have a selector. Why would we need to limit that behavior?
And if we do, why would we introduce a new selector for it? (instead of adding options to limit it on the existing `action` selector?)

**Reply by mdegat01** [reply]
The problem with the mobile app approach is it only allows you to notify your mobile app. What if I want to receive this notification via pushbullet? Telegram? HA itself (persistent notifications)? Any of the other myriad of integrations which support notifications?

Adding options to the action selector sounds good actually, limit the options that can be picked to only ones your blueprint understands. It will be a little painful for the blueprint author since they will have to merge the data collected in prior steps with the action object by treating it as a dictionary. But since templates can now return actual dictionary objects that's totally doable.

**Reply by frenck** [reply]
> The problem with the mobile app approach is it only allows you to notify your mobile app. What if I want to receive this notification via pushbullet? Telegram? HA itself (persistent notifications)?

That is more the notifications platform that might need some love and overhaul IMHO.

**Reply by mdegat01** [reply]
That's definitely a possibility. I can't really think of any examples to illustrate this problem other then notification services. That seems like a sign that it is a problem with notifications.

I guess the problem is today you can only describe devices in terms of what they are, their 'is-a' relationships in programming. This thing 'is-a' light, 'is-a' vacuum, 'is-a' Xiaomi Aqara Magic Cube, etc. But it doesn't really make sense to say a device 'is-a' notifier. It's not a notifier, it has the ability to capability to receive notifications. In programming I would make a notify interface and then have many different things implement it.

I don't really have any way to describe devices that way in HA. To define standardized and common capabilities and then reuse them across devices/services of a variety of different types. To say this phone/vacuum/light/whatever can also receive notifications in addition to all the stuff it can normally do. And then select anything which can receive notifications as part of a blueprint. It almost feels like Notification shouldn't really be a platform but... something else.

**Comment by frenck** [comment, marked as answer]
I'll go ahead and close this architectural proposal.

It was driven from the notification end, for which we now have an entity component. Migrating all notification services might take some time, but there is definitely a future paved for this one.

../Frenck
